src/System.Text.Primitives/System/Text/Encoders/Utf8Validator.cs

```diff
@@ -0,0 +1,252 @@
+        // Little-endian machine: [ PS3B, PS2B, PS1B, LEN ]
+        private uint _partialSequence;
+
+        private string DebuggerDisplay
+        {
+            get
+            {
+                if (_partialSequence == 0)
+                {
+                    return "Data VALID so far; no partial sequence consumed.";
+                }
+                else if (IsInvalid)
+                {
+                    return "Data INVALID.";
+                }
+                else
+                {
+                    switch ((byte)_partialSequence)
+                    {
+                        case 1:
+                            return (BitConverter.IsLittleEndian)
+                                ? FormattableString.Invariant($"Data VALID so far; partial sequence [ {(byte)(_partialSequence >> 8):X2} ] consumed.")
+                                : FormattableString.Invariant($"Data VALID so far; partial sequence [ {(_partialSequence >> 24):X2} ] consumed.");
+                        case 2:
+                            return (BitConverter.IsLittleEndian)
+                                ? FormattableString.Invariant($"Data VALID so far; partial sequence [ {(byte)(_partialSequence >> 8):X2} {(byte)(_partialSequence >> 16):X2} ] consumed.")
+                                : FormattableString.Invariant($"Data VALID so far; partial sequence [ {(_partialSequence >> 24):X2} {(_partialSequence >> 16):X2} ] consumed.");
+
+                        case 3:
+                            return (BitConverter.IsLittleEndian)
+                                ? FormattableString.Invariant($"Data VALID so far; partial sequence [ {(byte)(_partialSequence >> 8):X2} {(byte)(_partialSequence >> 16):X2} {(_partialSequence >> 24):X2} ] consumed.")
+                                : FormattableString.Invariant($"Data VALID so far; partial sequence [ {(_partialSequence >> 24):X2} {(_partialSequence >> 16):X2} {(byte)(_partialSequence >> 8):X2} ] consumed.");
+
+                        default:
+                            return "** INTERNAL ERROR **";
+                    }
+                }
+            }
+        }
+
+        private bool IsInvalid => IsInvalidPartialSequence(_partialSequence);
+
+        private unsafe static uint ConsumeDataWithExistingPartialSequence(uint partialSequence, ReadOnlySpan<byte> utf8Bytes)
+        {
+            Debug.Assert(partialSequence != 0 && !IsInvalidPartialSequence(partialSequence));
+
+            int partialSequenceOriginalByteCount = (byte)partialSequence;
+
+            if (BitConverter.IsLittleEndian)
+            {
+                // When we turn this into a Span<byte>, want MSB to be the first byte of the partial sequence
+                partialSequence >>= 8;
+            }
+
+            // Copy as much data as we can from the input buffer to our partial sequence.
+
+            Span<byte> partialSequenceAsBytes = new Span<byte>(&partialSequence, sizeof(uint));
```

Comment:
Can we use binary writer APIs instead of the unsafe constructor?
```C#
var span = new Span<byte>(new byte[4]);
WriteMachineEndian(span, value);
```

Reply:
Fixed.